kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/EncryptionDekCache.java

```diff
@@ -39,16 +41,60 @@ public class EncryptionDekCache<K, E> {
 
     public static final int NO_MAX_CACHE_SIZE = -1;
     private CipherSpecResolver cipherSpecResolver;
+    private AtomicLong invalidationCount = new AtomicLong(0);
 
     private record CacheKey<K>(K kek, CipherSpec cipherSpec) {}
 
+    public class EncryptionDek {
+        private final CacheKey<K> key;
+        private final Dek<E> dek;
+        private final AtomicBoolean hasBeenRotated = new AtomicBoolean(false);
+        private final CompletableFuture<Void> rotated = new CompletableFuture<>();
+
+        public EncryptionDek(CacheKey<K> key, Dek<E> dek) {
+            this.key = key;
+            this.dek = dek;
+        }
+
+        /**
+         * Destroy this DEK for encryption
+         * Discard the currently cached DEK for the KEK in the given {@code encryptionScheme} once.
+         * Competing threads will only discard it from the cache once.
+         * This method may block if a DEK for the given {@code encryptionScheme} is in the process
+         * of being loaded.
+         */
+        public void rotate() {
+            if (hasBeenRotated.compareAndSet(false, true)) {
+                LOGGER.debug("rotating encryption key {}", key);
+                dek.destroyForEncrypt();
+                EncryptionDekCache.this.invalidate(key);
+                rotated.complete(null);
```

Comment:
looks like we could chain to `var future = asyncCache.asMap().remove(key);` instead of the current `asyncCache.synchonous().invalidate(key);` style.

Reply:
IIRC `synchronous()` was in their for a reason can you double check the Caffine javadoc around `asMap` and `synchronous`?

Reply:
couldn't find anything in the [PR](https://github.com/kroxylicious/kroxylicious/pull/1001). this is where I pulled the [asMap from](https://github.com/ben-manes/caffeine/issues/422).

Reply:
```
  /**
   * Returns a view of the entries stored in this cache as a synchronous {@link LoadingCache}. A
   * mapping is not present if the value is currently being loaded. Modifications made to the
   * synchronous cache directly affect the asynchronous cache. If a modification is made to a
   * mapping that is currently loading, the operation blocks until the computation completes.
   *
   * @return a thread-safe synchronous view of this cache
   */
  @Override
  LoadingCache<K, V> synchronous();
```
vs
```
  /**
   * Returns a view of the entries stored in this cache as a thread-safe map. Modifications made to
   * the map directly affect the cache.
   * <p>
   * A computation operation, such as {@link ConcurrentMap#compute}, performs the entire method
   * invocation atomically, so the function is applied at most once per key. Some attempted update
   * operations by other threads may be blocked while computation is in progress. The computation
   * must not attempt to update any other mappings of this cache.
   * <p>
   * Iterators from the returned map are at least <i>weakly consistent</i>: they are safe for
   * concurrent use, but if the cache is modified (including by eviction) after the iterator is
   * created, it is undefined which of the changes (if any) will be reflected in that iterator.
   *
   * @return a thread-safe view of this cache supporting all of the optional {@link Map} operations
   */
  ConcurrentMap<K, CompletableFuture<V>> asMap();
```

Reply:
previously it would have needed to be synchronous to ensure it was removed from the cache before the re-attempt, but chaining onto the rotation future solves that.